kernel/src/engine/arrow_expression.rs

```diff
@@ -350,6 +352,190 @@ fn evaluate_expression(
+fn apply_schema_to_struct(array: &dyn Array, kernel_fields: &Schema) -> DeltaResult<StructArray> {
+    let Some(sa) = array.as_struct_opt() else {
+        return Err(make_arrow_error(
+            "Arrow claimed to be a struct but isn't a StructArray",
+        ));
+    };
+    transform_struct(sa, kernel_fields.fields())
+}
+
+// deconstruct the array, then rebuild the mapped version
+fn apply_schema_to_list(
+    array: &dyn Array,
+    target_inner_type: &ArrayType,
+) -> DeltaResult<ListArray> {
+    let Some(la) = array.as_list_opt() else {
+        return Err(make_arrow_error(
+            "Arrow claimed to be a list but isn't a ListArray",
+        ));
+    };
+    let (field, offset_buffer, values, nulls) = la.clone().into_parts();
+    let (transformed_field, transformed_values) = transform_field_and_col(
+        &field,
+        values,
+        &target_inner_type.element_type,
+        target_inner_type.contains_null,
+        None,
+        None,
+    )?;
+    Ok(ListArray::try_new(
+        Arc::new(transformed_field),
+        offset_buffer,
+        transformed_values,
+        nulls,
+    )?)
+}
+
+// deconstruct a map, and rebuild it with the specified target kernel type
+fn apply_schema_to_map(array: &dyn Array, kernel_map_type: &MapType) -> DeltaResult<MapArray> {
+    let Some(ma) = array.as_map_opt() else {
+        return Err(make_arrow_error(
+            "Arrow claimed to be a map but isn't a MapArray",
+        ));
+    };
+
+    let ArrowDataType::Map(arrow_map_type, _) = array.data_type() else {
+        return Err(make_arrow_error(
+            "Arrow claimed to be a map but doesn't have map DataType",
+        ));
+    };
+
+    let (map_field, offset_buffer, map_struct_array, nulls, ordered) = ma.clone().into_parts();
+
+    let ArrowDataType::Struct(_) = arrow_map_type.data_type() else {
+        return Err(make_arrow_error("Arrow map type wasn't a struct."));
+    };
+
+    let target_fields = map_struct_array
+        .fields()
+        .iter()
+        .zip([
+            (&kernel_map_type.key_type, false),
+            (
+                &kernel_map_type.value_type,
+                kernel_map_type.value_contains_null,
+            ),
+        ])
+        .map(|(arrow_field, (target_type, nullable))| {
+            StructField::new(arrow_field.name(), target_type.clone(), nullable)
+        });
+
+    // Arrow puts the key type/val as the first field/col and the value type/val as the second. So
+    // we just transform like a 'normal' struct, but we know there are two fields/cols and we
+    // specify the key/value types as the target type iterator.
+    let transformed_map_struct_array = transform_struct(&map_struct_array, target_fields)?;
+
+    let transformed_map_field = Arc::new(
+        map_field
+            .as_ref()
+            .clone()
+            .with_data_type(transformed_map_struct_array.data_type().clone()),
+    );
+    Ok(MapArray::try_new(
+        transformed_map_field,
```

Comment:
nit: cheating `cargo fmt`...
```suggestion
    let transformed_map_field = map_field
        .as_ref()
        .clone()
        .with_data_type(transformed_map_struct_array.data_type().clone());
    Ok(MapArray::try_new(
        Arc::new(transformed_map_field),
```